### DISCUSSION TITLE
ILP TCP randomly dropping rows

### DISCUSSION THREAD

**Discussion by robocoder** [questdb/questdb#1291] [Ideas]
I'm testing out QuestDB 6.0.4 (default config). Using the same data each time (~1M rows), I'm finding different rows are randomly dropped each time. The errors are always of the variety:

```
TIMESTAMP E i.q.c.l.t.LineTcpConnectionContext [29] could not parse measurement, code INVALID_FIELD_SEPERATOR at XXX line (may be mangled due to partial parsing) is table,symbolList fieldValueList timestamp
```

(The line number reported also varies, e.g., 149, 155, 306, 312, 322, ...)

There's nothing wrong with the rejected rows. I can re-ingest them without error. Is this considered acceptable behaviour, or is it a bug?

**Comment by bsmth** [comment]
Hi @robocoder, thanks for the discussion I've opened an issue to track this one here -> https://github.com/questdb/questdb/issues/1346

**Comment by ideoma** [comment]
Hey @robocoder,

Can you please post the piece of code you use to send the data? It can be that the client socket does not send the whole buffer before you continue to the next one.

The data example can also be good.

**Reply by robocoder** [reply]
On Linux, php 8.0's `fwrite` is a wrapper that takes care of all the buffering/chunking.

```php
$fp = fsockopen('localhost', 9009, $errno, $errstr, 10);
fwrite($fp, $data);
fclose($fp);
```

**Reply by ideoma** [reply]
If it's PHP then perhaps it's this 

https://stackoverflow.com/questions/66145076/php-fwrite-does-not-write-to-socket

Also this seems relevant

https://githubmemory.com/repo/metaregistrar/php-epp-client/issues/263

**Comment by marregui** [comment]
Hi @robocoder 

Could I please have some of your data to run some local tests?

Thank you in advance

**Reply by robocoder** [reply]
I'll have to get permission and likely anonymize the contents. I'll look into it.

**Reply by robocoder** [reply]
Ok, it's still happening with questdb 6.0.7.1.

Attempting to import 1082397 rows via ILP TCP. 6 runs dropping 6, 1, 2, 3, 3, and 3 rows on the corresponding runs.

@ideoma I modified the code to handle `fwrite()` returning a count less than the expected number of bytes. PHP isn't the problem.

I'll try to provide a data file this weekend.

**Comment by bluestreak01** [comment]
We're working on hardening ILP parsers, should have something ready next week

**Comment by bluestreak01** [comment]
@robocoder We have made a number of fixes to the TCP ILP parser. Would you like us to cut you a version to try on your data ?

**Reply by robocoder** [reply]
sure, I can try it out.

**Reply by robocoder** [reply]
The issue hasn't been resolved with 6.0.8.  I now get "Connection reset by peer".

The server log is:
```
2021-10-15T00:54:48.494373Z E i.q.c.l.t.LineTcpConnectionContext [53] could not process line data [table=foo, ex=^M
java.lang.InternalError: a fault occurred in a recent unsafe memory access operation in compiled Java code^M
    at io.questdb.std.Files.mmap(Files.java:181)^M
    at io.questdb.std.Files.mmap(Files.java:176)^M
    at io.questdb.std.FilesFacadeImpl.mmap(FilesFacadeImpl.java:179)^M
    at io.questdb.cairo.TableUtils.mapRO(TableUtils.java:403)^M
    at io.questdb.cairo.TableUtils.mapRO(TableUtils.java:399)^M
    at io.questdb.cairo.vm.MemoryCMRImpl.setSize0(MemoryCMRImpl.java:111)^M
    at io.questdb.cairo.vm.MemoryCMRImpl.extend(MemoryCMRImpl.java:67)^M
    at io.questdb.cairo.SymbolMapReaderImpl.growCharMemToSymbolCount(SymbolMapReaderImpl.java:216)^M
    at io.questdb.cairo.SymbolMapReaderImpl.updateSymbolCount(SymbolMapReaderImpl.java:89)^M
    at io.questdb.cutlass.line.tcp.SymbolCache.getSymIndex(SymbolCache.java:68)^M
    at io.questdb.cutlass.line.tcp.LineTcpMeasurementScheduler$TableUpdateDetails$ThreadLocalDetails.getSymbolIndex(LineTcpMeasurementScheduler.java:1196)^M
    at io.questdb.cutlass.line.tcp.LineTcpMeasurementScheduler$TableUpdateDetails.getSymbolIndex(LineTcpMeasurementScheduler.java:1018)^M
    at io.questdb.cutlass.line.tcp.LineTcpMeasurementScheduler$LineTcpMeasurementEvent.createMeasurementEvent(LineTcpMeasurementScheduler.java:517)^M
    at io.questdb.cutlass.line.tcp.LineTcpMeasurementScheduler.tryButCouldNotCommit(LineTcpMeasurementScheduler.java:401)^M
    at io.questdb.cutlass.line.tcp.LineTcpConnectionContext.parseMeasurements(LineTcpConnectionContext.java:175)^M
    at io.questdb.cutlass.line.tcp.LineTcpConnectionContext.handleIO(LineTcpConnectionContext.java:158)^M
    at io.questdb.cutlass.line.tcp.LineTcpMeasurementScheduler$NetworkIOJobImpl.handleIO(LineTcpMeasurementScheduler.java:1490)^M
    at io.questdb.cutlass.line.tcp.LineTcpMeasurementScheduler$NetworkIOJobImpl.onRequest(LineTcpMeasurementScheduler.java:1508)^M
    at io.questdb.network.AbstractIODispatcher.processIOQueue(AbstractIODispatcher.java:162)^M
    at io.questdb.cutlass.line.tcp.LineTcpMeasurementScheduler$NetworkIOJobImpl.run(LineTcpMeasurementScheduler.java:1438)^M
    at io.questdb.mp.Worker.run(Worker.java:112)^M
]^M
2021-10-15T00:54:48.495018Z I tcp-line-server scheduling disconnect [fd=53, reason=0]^M
2021-10-15T00:54:48.495087Z I tcp-line-server disconnected [ip=127.0.0.1, fd=53, src=queue]^M
```

**Reply by bluestreak01** [reply]
Is this a new table or did it exist before 6.0.8?

**Reply by robocoder** [reply]
for my benchmarking, it's usually an existing table that I truncate between runs. but I still encountered this problem after I dropped and recreated it.

**Reply by bluestreak01** [reply]
We managed to reproduce this, thanks!

**Comment by robocoder** [comment]
I noticed that if the client reconnects (after getting the `Connection reset by peer` error), the table count is corrupted (in addition to dropping rows).

```
$s = socket_create(AF_INET, SOCK_STREAM, SOL_TCP);
socket_connect($s, 'localhost', 9009);

foreach ($rows as $row) {
    try {
        socket_write($s, $row);
    } catch(\Exception $e) {
        // reconnect and re-attempt
        $s = socket_create(AF_INET, SOCK_STREAM, SOL_TCP);
        socket_connect($s, 'localhost', 9009);
        socket_write($s, $row);
    }
}

socket_close($s);
```

Number of rows being written: 4673981

QuestDB count suggests more rows than were written:
```
admin=> select count() from foo;
  count  
---------
 5004639
(1 row)
```

But a table scan says otherwise:
```
admin=> select count() from foo where email != 'robocoder@example.com';
  count  
---------
 4673825
(1 row)

admin=> select count() from foo where email = 'robocoder@example.com';
 count 
-------
     0
(1 row)
```

**Reply by bluestreak01** [reply]
Thank you Anthon, is “email” field an indexed symbol? If so it is index that could be corrupted rather than table. Index can be rebuilt easily.

We do have a fix for the “connection reset” problem. Should be able to release on Monday

**Reply by robocoder** [reply]
"email" is an unindexed string field. Also, the count (4673825) jives with the results of `select count() from foo sample by 1d;` (when summed up).

**Comment by bluestreak01** [comment]
I am looking into the count discrepancy in the mean time.

**Reply by bluestreak01** [reply]
I am failing to reproduce the error with incorrect row counts. Having said this, the original problem is to do with random memory corruption. We fixed the original issue here: https://github.com/questdb/questdb/pull/1441 as well as described what's happening if this is of interest.

**Reply by robocoder** [reply]
ok. let's hope the mis-count is just a side effect.

**Reply by robocoder** [reply]
Ran new tests against 6.0.9. The unconstrained count() is off. I'm expecting 5414744 rows:
```
admin=> select count() from foo;
  count  
---------
 5678988
(1 row)

admin=> select sum(count) from (select count() as "count" from foo sample by 1d);
   sum   
---------
 5414744
(1 row)
```

I'm going to export the table to see if everything was ingested properly.

**Reply by robocoder** [reply]
I guess there's some background processing still running... the count () changed:
```
admin=> select count() from foo;
  count  
---------
 5950677
(1 row)

admin=> select sum(count) from  (select count() as "count" from foo sample by 1d);
   sum   
---------
 5414744
(1 row)
```

**Reply by robocoder** [reply]
Exported tables, previously imported via REST and ILP-TCP, both match up 100%, despite errors in the log (along the lines of java.lang.InternalError and java.lang.AssertionError).

(edit: I'm actually not sure if the disconnects are gone as I forgot I had reconnect logic in my client).

Apologies for the delay in providing sample data...

**Reply by ideoma** [reply]
Thanks for having a look.

Can you post errors from the log please?

**Reply by bluestreak01** [reply]
I wonder if this is test client can we perhaps have it along with the data you publish so we can fix everything at once?

**Comment by bluestreak01** [comment]
We just released 6.1, which should fix all of these issues. Do you mind re-running the failing process again please ?

**Reply by robocoder** [reply]
Thanks for the heads-up. I only see source tarballs in github. Will you be releasing platform packages?

**Comment by bluestreak01** [comment]
yeah, will be release today. There is docker image or you could build from source (master) if you like:

```
mvn clean package --batch-mode --quiet -DskipTests -P build-web-console,build-binaries
```

**Comment by robocoder** [comment]
Thanks. 6.1.1 has fixed issues (#1346 and #1459) with my data.

**Reply by bluestreak01** [reply]
Awesome, thank you for confirming 👍
